dvc/remote/base.py

```diff
@@ -389,188 +530,400 @@ def supported(cls, config):
+                traverse_prefixes += [
+                    "{0:0{1}x}".format(i, self.TRAVERSE_PREFIX_LEN)
+                    for i in range(1, pow(16, self.TRAVERSE_PREFIX_LEN - 2))
+                ]
+        with Tqdm(
+            desc="Querying "
+            + (f"cache in '{name}'" if name else "remote cache"),
+            total=remote_size,
+            initial=initial,
+            unit="file",
+        ) as pbar:
 
-        if checksum:
-            self.state.save(path_info, checksum)
+            def list_with_update(prefix):
+                return list(
+                    self.list_checksums(
+                        prefix=prefix, progress_callback=pbar.update
+                    )
+                )
 
-        return checksum
+            with ThreadPoolExecutor(max_workers=jobs or self.JOBS) as executor:
+                in_remote = executor.map(list_with_update, traverse_prefixes,)
+                yield from itertools.chain.from_iterable(in_remote)
+
+    def _list_checksums_exists(self, checksums, jobs=None, name=None):
+        logger.debug(
+            "Querying {} checksums via object_exists".format(len(checksums))
+        )
+        with Tqdm(
+            desc="Querying "
+            + ("cache in " + name if name else "remote cache"),
+            total=len(checksums),
+            unit="file",
+        ) as pbar:
 
-    def save_info(self, path_info):
-        return {self.PARAM_CHECKSUM: self.get_checksum(path_info)}
+            def exists_with_progress(path_info):
+                ret = self.tree.exists(path_info)
+                pbar.update_msg(str(path_info))
+                return ret
+
+            with ThreadPoolExecutor(max_workers=jobs or self.JOBS) as executor:
+                path_infos = map(self.checksum_to_path_info, checksums)
+                in_remote = executor.map(exists_with_progress, path_infos)
+                ret = list(itertools.compress(checksums, in_remote))
+                return ret
+
+    @index_locked
+    def gc(self, named_cache, jobs=None):
+        used = set(named_cache.scheme_keys("local"))
+
+        if self.scheme != "":
+            used.update(named_cache.scheme_keys(self.scheme))
+
+        removed = False
+        # checksums must be sorted to ensure we always remove .dir files first
+        for checksum in sorted(
+            self.all(jobs, str(self.path_info)),
+            key=self.is_dir_checksum,
+            reverse=True,
+        ):
+            if checksum in used:
+                continue
+            path_info = self.checksum_to_path_info(checksum)
+            if self.is_dir_checksum(checksum):
+                # backward compatibility
+                self._remove_unpacked_dir(checksum)
+            self.tree.remove(path_info)
+            removed = True
+        if removed:
+            self.index.clear()
+        return removed
+
+    def _remove_unpacked_dir(self, checksum):
+        pass
+
+
+class CacheMixin:
```

Comment:
Cloud cache is technically not an extension to cloud remote, since we define remote as the cloud push/pull mirror for a local .dvc MD5 checksum-based cache. However, cloud caches still need to implement the remote methods for size estimation and garbage collection (so that we can display progress during `gc`).

Originally, I had separate `BaseCloud`, `RemoteMixin`, `CacheMixin`, classes, but it seemed strange to have it like that if every cache class was always going to include the `RemoteMixin` methods. So for now all the cloud cache classes are implemented as an extension of a cloud remote class + the cache mixin.